src/main/java/com/google/error_analyzer/backend/Pagination.java

```diff
@@ -0,0 +1,211 @@
+        int page = Integer.parseInt(request.getParameter("requestedPage"));
+        String fileName = request.getParameter("fileName");
+        String fileType = request.getParameter("fileType");
+        String next = request.getParameter("next");
+
+        SearchErrors searchErrors = new SearchErrors();
+        HashMap < String, String > search = searchErrors.getSearchedErrors();
+        if (page <= 0)
+            page = 1;
+        // if file field is empty the user haven't choosen anyfile so return empty object
+        if (fileName.length() == 0) {
+            returnEmptyObject(response);
+            return;
+        }
+
+        // if fileName is not empty fetch the data from database and return response
+        fetchAndReturnResponse(page, fileName, fileType, next, search, response);
+
+    }
+
+    //  if user is asking for 1st page return that page and make a window of fully finished and ready to use data for that file
+    // if user is not on 1st page but someother page just maintain the already created window of pages
+
+    public void fetchAndReturnResponse(int page, String fileName, String fileType, String next, HashMap < String, String > search, HttpServletResponse response) throws IOException {
+        if (page == 1) {
+            // the user won't be knowing the name of error file generated corresponding to his log file so we need to change fileName depending on type of file
+            if (fileType.equals(ERROR))
+                fileName = fileName + "error";
+            lastPage=Integer.MAX_VALUE;
+            // create window
+            fetchData(0, (recordsPerPage * noOfPages) - 1, 0, fileName,fileType, page, search);
+            returnResponse(0, recordsPerPage - 1, isLastPage(page), response);
+        } else {
+            int start = recordsPerPage * ((page - 1) % noOfPages);
+            int stop = start + recordsPerPage - 1;
+            if (isLastPage(page))
+                stop = start + noOfRecordsOnLastPage - 1;
+            returnResponse(start, stop, isLastPage(page), response);
+            // maintain window
+            maintainWindow(page, next, fileName,fileType, search);
+        }
+    }
+    
+
+    // returns true if the asked page is last page
+    public boolean isLastPage(int page){
+        return page == lastPage ? true : false;
+    }
+
+    // picks content from maintained window on the basis of required page and convert it into json format
+    private void returnResponse(int startIdx, int stopIdx, boolean isLastPage, HttpServletResponse response) throws IOException {
+        ArrayList < String > display = new ArrayList();
+        for (int i = startIdx; i <= stopIdx && i < data.size(); i++) {
+            display.add(data.get(i));
+        }
+        
+        String json = convertToJson(new logOrErrorResponse(display, isLastPage));
+        response.setContentType("application/json");
```

Comment:
Only doPost and doGet should write to response. 
These helper functions should only return appropriate results back to the caller function.